---
title: "scRNAseq import and clustering"
author: "Miranda Hunter"
date: "5/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries:
``` {r, message = F, warning = F}
library(Seurat)
library(tidyverse)
library(scales)
library(pals)
library(cowplot)
library(fgsea)
library(msigdbr)
library(data.table)
library(mvhfunctions) # https://github.com/mvhunter1/mvhfunctions
```


Import counts matrices and create Seurat objects.
``` {r}
groups <- c("XFG", "XFN", "XBG", "XBN")

for (group in groups) {
  path <- paste0('/Users/hunterm/Dropbox/MH_ST/Miranda_R/Josh_singlecell/cellranger_h5/', group, '/filtered_feature_bc_matrix/')
  counts <- Read10X(path)
  seurat_obj <- CreateSeuratObject(counts = counts,
                                   project = group,
                                   min.cells = 3,
                                   min.features = 200)
  assign(group, seurat_obj)
  rm(path, counts, seurat_obj)
}

```


Look at some QC metrics and add to metadata.
Also normalize, scale data and perform PCA.
``` {r}

for (group in groups) {
  seurat_obj <- eval(parse(text = group))
  
  # add some QC metrics to metadata
  seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^mt-")
  ribo_genes <- grep(pattern = "^rps|^rpl", x = rownames(seurat_obj), value = T)
  seurat_obj[["percent.ribosomal"]] <- PercentageFeatureSet(seurat_obj, features = ribo_genes)
  
  seurat_obj <- SCTransform(seurat_obj, verbose = F) # normalization
  seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", verbose = F)
  seurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj), verbose = F)
  seurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj), verbose = F)
  
  assign(group, seurat_obj)
  rm(seurat_obj)
}
```


Determine dimensionality.
``` {r}
ElbowPlot(XBG) + ggtitle('XBG')
ElbowPlot(XBN) + ggtitle('XBN')
ElbowPlot(XFG) + ggtitle('XFG')
ElbowPlot(XFN) + ggtitle('XFN')
```
Will go with 15 PCs for each because it seems pretty consistent.


Run UMAP and clustering.
``` {r}
for (group in groups) {
  seurat_obj <- eval(parse(text = group))
  
  seurat_obj <- FindNeighbors(seurat_obj, dims = 1:15, verbose = F)
  seurat_obj <- FindClusters(seurat_obj, resolution = 0.8, verbose = F)
  seurat_obj <- RunUMAP(seurat_obj, dims = 1:15, verbose = F)
  assign(group, seurat_obj)
  rm(seurat_obj)
}
```


Some plots.
``` {r}
DimPlot(XBG) + ggtitle('XBG')
DimPlot(XBN) + ggtitle('XBN')
DimPlot(XFG) + ggtitle('XFG')
DimPlot(XFN) + ggtitle('XFN')
```


Plot transgene expression.
``` {r}
p1 <- nice_feature_plot(XFG, features = "hsCRKL", order = T, pt.size = 0.5) + ggtitle('XFG hsCRKL')
p2 <- nice_feature_plot(XFN, features = "hsCRKL", order = T, pt.size = 0.5) + ggtitle('XFN hsCRKL')
p3 <- nice_feature_plot(XBG, features = "hsCRKL", order = T, pt.size = 0.5) + ggtitle('XBG hsCRKL')
p4 <- nice_feature_plot(XBN, features = "hsCRKL", order = T, pt.size = 0.5) + ggtitle('XBN hsCRKL')
plot_grid(p1,p2,p3,p4, nrow = 2)
```

``` {r}
p1 <- nice_feature_plot(XFG, features = "hsGAB2", order = T, pt.size = 0.5) + ggtitle('XFG hsCRKL')
p2 <- nice_feature_plot(XFN, features = "hsGAB2", order = T, pt.size = 0.5) + ggtitle('XFN hsCRKL')
p3 <- nice_feature_plot(XBG, features = "hsGAB2", order = T, pt.size = 0.5) + ggtitle('XBG hsCRKL')
p4 <- nice_feature_plot(XBN, features = "hsGAB2", order = T, pt.size = 0.5) + ggtitle('XBN hsCRKL')
plot_grid(p1,p2,p3,p4, nrow = 2)
```


Merge the 4 objects into 1.
Need to merge Seurat objects in a stepwise way - doing all 4 at once causes an out of memory error.
``` {r}
options(future.globals.maxSize = 8000 * 1024^2) # increase memory

body <- merge(XBG, y = XBN, add.cell.ids = c("XBG", "XBN"))
fin <- merge(XFG, y = XFN, add.cell.ids = c("XFG", "XFN"))

XYZ_all <- merge(body, y = fin)
```


Rerun normalization and clustering on the new object.
``` {r, message = F, warning = F}
XYZ_all <- SCTransform(XYZ_all, verbose = F)
XYZ_all <- FindVariableFeatures(XYZ_all, selection.method = "vst", verbose = F)
XYZ_all <- RunPCA(XYZ_all, features = VariableFeatures(XYZ_all), verbose = F)
ElbowPlot(XYZ_all)
```
Will choose 15 PCs for dimensionality reduction.


Clustering and UMAP.
``` {r}
XYZ_all <- FindNeighbors(XYZ_all, dims = 1:15, verbose = F)
XYZ_all <- FindClusters(XYZ_all, resolution = 0.8, verbose = F)
XYZ_all <- RunUMAP(XYZ_all, dims = 1:15, verbose = F)
DimPlot(XYZ_all)
DimPlot(XYZ_all, group.by = "orig.ident")
```


Refine cluster resolution.
``` {r}
XYZ_all <- FindClusters(XYZ_all, resolution = 0.8, verbose = F)
XYZ_all <- FindClusters(XYZ_all, resolution = 0.5, verbose = F)
XYZ_all <- FindClusters(XYZ_all, resolution = 0.4, verbose = F)
XYZ_all <- FindClusters(XYZ_all, resolution = 0.3, verbose = F)
XYZ_all <- FindClusters(XYZ_all, resolution = 0.2, verbose = F)


DimPlot(XYZ_all, group.by = "SCT_snn_res.0.8", label = T)
DimPlot(XYZ_all, group.by = "SCT_snn_res.0.5", label = T)
DimPlot(XYZ_all, group.by = "SCT_snn_res.0.4", label = T)
DimPlot(XYZ_all, group.by = "SCT_snn_res.0.3", label = T) 
DimPlot(XYZ_all, group.by = "SCT_snn_res.0.2", label = T) # will choose this resolution
```


Marker gene scoring:
``` {r}
cancer_markers <- c("BRAFhuman", "crestin", "sox10", "mitfa", "tyrp1a", "tyrp1b", "dct", "pmela")
ker_markers <- c("tp63", "krt4", "krt15", "krt17", "krt18")
fibro_markers <- c("fn1a", "colla1a", "fstl1b", "s100a10b")
eryth_markers <- c("ba1", "cahz", "hbaa1")
nk_markers <- c("nkl.2", "ccl38.6")
neutr_markers <- c("mpx", "mmp9", "ncf1")
macro_markers <- c("ctss2.1", "ctss2.2", "mhc2dab")

markers_all <- list(tumor_markers = cancer_markers, 
                    ker_markers = ker_markers, 
                    fibro_markers = fibro_markers, 
                    eryth_markers = eryth_markers, 
                    nk_markers = nk_markers, 
                    neutr_markers = neutr_markers, 
                    macro_markers = macro_markers)


for (ii in 1:length(markers_all)) {
  marker_genes <- intersect(markers_all[[ii]], rownames(XYZ_all))
  marker_name <- names(markers_all[ii])
  XYZ_all <- AddModuleScore(XYZ_all,
                            features = list(marker_genes),
                            name = marker_name)
}

nice_feature_plot(XYZ_all, features = "tumor_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "ker_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "fibro_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "eryth_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "nk_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "neutr_markers1", order = T, pt.size = 0.5)
nice_feature_plot(XYZ_all, features = "macro_markers1", order = T, pt.size = 0.5)
nice_dim_plot(XYZ_all, group_by = "SCT_snn_res.0.2")
```


Find markers:
``` {r}
Idents(XYZ_all) <- "SCT_snn_res.0.2"
XYZ_markers_all <- FindAllMarkers(XYZ_all)
```

Cluster annotations:
5 = xanthophores
10,7,1,0 = melanocytes
16 = interface
2,4 = erythrocytes
3,13 = NK cells
8 = keratinocytes
6 = fibroblasts
9 = macrophages
12 = neutrophils
11 = T cells
14 = unclear
15 = muscle-like

Add cluster IDs to metadata.
``` {r}
XYZ_metadata <- XYZ_all[[]] %>% rownames_to_column(var = "cell") %>% dplyr::select(cell, SCT_snn_res.0.2)
for (ii in 1:nrow(XYZ_metadata)) {
  cluster <- XYZ_metadata[ii,2] %>% as.character() %>% as.numeric()
  if (cluster == 5) {
    XYZ_metadata$cell_type[ii] <- "xanthophores"
  } else if (cluster %in% c(0,1,7,10)) {
    XYZ_metadata$cell_type[ii] <- "melanocytes"
  } else if (cluster %in% c(2,4)) {
    XYZ_metadata$cell_type[ii] <- "erythrocytes"
  } else if (cluster == 16) {
    XYZ_metadata$cell_type[ii] <- "interface"
  } else if (cluster %in% c(3,13)) {
    XYZ_metadata$cell_type[ii] <- "NK cells"
  } else if (cluster == 8) {
    XYZ_metadata$cell_type[ii] <- "keratinocytes"
  } else if (cluster == 6) {
    XYZ_metadata$cell_type[ii] <- "fibroblasts"
  } else if (cluster == 9) {
    XYZ_metadata$cell_type[ii] <- "macrophages"
  } else if (cluster == 12) {
    XYZ_metadata$cell_type[ii] <- "neutrophils"
  } else if (cluster == 11) {
    XYZ_metadata$cell_type[ii] <- "T cells"
  } else if (cluster == 14) {
    XYZ_metadata$cell_type[ii] <- "unknown"
  } else if (cluster == 15) {
    XYZ_metadata$cell_type[ii] <- "muscle-like"
  }
}

XYZ_all <- AddMetaData(XYZ_all,
                   metadata = XYZ_metadata %>% dplyr::select(cell, cell_type) %>% deframe(),
                   col.name = "cell_type")

nice_dim_plot(XYZ_all, group_by = "cell_type")
```


Make melanocyte-only dataset.
``` {r}
Idents(XYZ_all) <- "cell_type"
all_mels <- subset(XYZ_all, idents = "melanocytes")
all_mels

# remove old cluster assignment
SCT_snn_res.0.5 <- NULL
SCT_snn_res.0.4 <- NULL
SCT_snn_res.0.45 <- NULL
SCT_snn_res.0.3 <- NULL
SCT_snn_res.0.2 <- NULL

all_mels <- FindNeighbors(all_mels, dims = 1:15, verbose = F)
all_mels <- RunUMAP(all_mels, dims = 1:15, verbose = F)
all_mels <- FindClusters(all_mels, resolution = 0.3, verbose = F)
all_mels <- FindClusters(all_mels, resolution = 0.2, verbose = F)
all_mels <- FindClusters(all_mels, resolution = 0.1, verbose = F)
nice_dim_plot(all_mels, group_by = "SCT_snn_res.0.3")
nice_dim_plot(all_mels, group_by = "SCT_snn_res.0.2")
nice_dim_plot(all_mels, group_by = "SCT_snn_res.0.1")
nice_dim_plot(all_mels, group_by = "orig.ident", pt_size = 0.2, label = F)
```


## Plot average expression of fin melanocyte hox genes.
``` {r}
fin_genes <- c("hoxa13a", "hoxa13b", "hoxb13a", "hoxc13a", "hoxc13b")
body_genes <- c("hoxb5a", "hoxb6a", "hoxb7a", "hoxb8a", "hoxb9a")

# extract normalized data from Seurat object
fin_expression <- GetAssayData(all_mels, slot = "data") %>% .[rownames(.) %in% fin_genes,] %>% t()
body_expression <- GetAssayData(all_mels, slot = "data") %>% .[rownames(.) %in% body_genes,] %>% t()

# calculate average expression for fin vs body genes
fin_expression <- fin_expression %>% data.frame() %>% mutate(mean_exp = rowMeans(.)) %>% rownames_to_column(var = "cell")
body_expression <- body_expression %>% data.frame() %>% mutate(mean_exp = rowMeans(.)) %>% rownames_to_column(var = "cell")

# remove anything that is XFN or XBN (aka not melanocytes)
fin_expression <- fin_expression %>% .[grep(pattern = "XBN|XFN", x = fin_expression$cell, invert = T),]
body_expression <- body_expression %>% .[grep(pattern = "XBN|XFN", x = body_expression$cell, invert = T),]


# add another column saying whether melanocyte or fin
for (ii in 1:nrow(fin_expression)) {
  group <- fin_expression$cell[ii] %>% substr(., 1, 3)
  
  if (group == "XBG") {
    fin_expression$group[ii] <- "body"
  } else if (group == "XFG") {
    fin_expression$group[ii] <- "fin"
  }
}

for (ii in 1:nrow(body_expression)) {
  group <- body_expression$cell[ii] %>% substr(., 1, 3)
  
  if (group == "XBG") {
    body_expression$group[ii] <- "body"
  } else if (group == "XFG") {
    body_expression$group[ii] <- "fin"
  }
}

# fin expression
ggplot(fin_expression, aes(x = group, y = mean_exp, fill = group)) +
  geom_violin() +
  geom_jitter() +
  theme_classic() +
  theme(axis.title.x = element_blank(),
        axis.text = element_text(size = 14, color = "black"),
        axis.title.y = element_text(size = 16, color = "black", margin = margin(r = 10)),
        legend.position = "none") +
  ylab("average expression of\nhox13 genes")

# body expression
ggplot(body_expression, aes(x = group, y = mean_exp)) +
  geom_violin() +
  geom_jitter() +
  theme_classic() +
  theme(axis.title.x = element_blank(),
        axis.text = element_text(size = 14, color = "black"),
        axis.title.y = element_text(size = 16, color = "black", margin = margin(r = 10)),
        legend.position = "none") +
  ylab("average expression of\nbody hox genes")

```


Perform stats on fin expression data.
``` {r}
fin_cells <- fin_expression %>% filter(group == "fin")
body_cells <- fin_expression %>% filter(group == "body")
wilcox.test(fin_cells$mean_exp, body_cells$mean_exp)
fin_stats <- wilcox.test(fin_cells$mean_exp, body_cells$mean_exp)
fin_stats$p.value
```


Plot IGF gene expression.
``` {r}
igf_genes <- c("igf1", "igf2a", "igf2b")

# extract normalized data from Seurat object
igf_expression <- GetAssayData(all_mels, slot = "data") %>% .[rownames(.) %in% igf_genes,] %>% t()

# calculate average expression for fin vs body genes
igf_expression <- igf_expression %>% data.frame() %>% mutate(mean_exp = rowMeans(.)) %>% rownames_to_column(var = "cell")

# remove anything that is XFN
igf_expression <- igf_expression %>% .[grep(pattern = "XBN|XFN", x = igf_expression$cell, invert = T),]


# add another column saying whether melanocyte or fin
for (ii in 1:nrow(igf_expression)) {
  group <- igf_expression$cell[ii] %>% substr(., 1, 3)
  
  if (group == "XBG") {
    igf_expression$group[ii] <- "body"
  } else if (group == "XFG") {
    igf_expression$group[ii] <- "fin"
  }
}

ggplot(igf_expression, aes(x = group, y = mean_exp, fill = group)) +
  geom_violin() +
  geom_jitter() +
  theme_classic() +
  theme(axis.title.x = element_blank(),
        axis.text = element_text(size = 14, color = "black"),
        axis.title.y = element_text(size = 16, color = "black", margin = margin(r = 10)),
        legend.position = "none") +
  ylab("average expression of IGF genes")
```
